Add `start_line` and `end_line` information in `linkcode` (#9625)
sphinx-doc/sphinx: doc/usage/extensions/linkcode.rst

   - ``javascript``: ``object`` (name of the object), ``fullname``
     (name of the item)

>    The ``start_line`` and ``end_line`` keys are also always present in the
>    ``info`` dictionnary, but may be None if they are not found. By default,
>    theses values are found using the :event:`viewcode-find-source` event but you

Reviewer: ```suggestion
   The ``start_line`` and ``end_line`` keys are always present in the
   ``info`` dictionary. If start and end line numbers are not found, the keys
   have the default value ``-1``.
   By default, these values are found using the :event:`viewcode-find-source` event but you
```

          if not info['module']:
              return None
          filename = info['module'].replace('.', '/')
          line_anchor = ""

>           if info['start_line'] is not None:

Reviewer: ```suggestion
          if info['start_line'] >= 0:
```

   using the same import mechanism used by :mod:`sphinx.ext.viewcode`
   (:class:`~sphinx.pycode.ModuleAnalyzer`) after the
   :event:`viewcode-find-source` event has been tried. Otherwise only the event
   is used.

   .. warning::

      The same warning as in :mod:`sphinx.ext.viewcode` applies: using this
      method will import the modules being linked to.
      If any modules have side effects on import, these will be executed when
      ``sphinx-build`` is run.

      If you document scripts (as opposed to library modules), make sure their
      main routine is protected by a ``if __name__ == '__main__'`` condition.

   The default is ``False``.

.. event:: viewcode-find-source (app, modname)

   .. versionadded:: 4.1.3

   Find the source code for a module.
   An event handler for this event should return a tuple of the source code
   itself and a dictionary of tags.
   The dictionary maps the name of a class, function, attribute, etc to a tuple
   of its type, the start line number, and the end line number.
   The type should be one of "class", "def", or "other".

   :param app: The Sphinx application object.

>    :param modname: The name of the module to find source code for.

Reviewer: ```suggestion
```

          if not info['module']:
              return None
          filename = info['module'].replace('.', '/')
          line_anchor = ""
          if info['start_line'] >= 0:

>             line_anchor = "#L%s" % info['start_line']

Reviewer 2: Any plans to add the `end_line` to the line anchor?  It seems that the format `#L10-L23` is supported by a few web-based code viewers (checked: [GitHub](https://github.com/sphinx-doc/sphinx/blob/faa33a53a389f6f8bc1f6ae97d6015fa92393c4a/README.rst?plain=1#L1-L3), [GitLab](https://gitlab.com/gitlab-org/gitlab/-/blob/23451dc906d3d43a2aca49ec606d01d39363ec4e/.gitlab-ci.yml#L21-24), [Gitea](https://gitea.com/gitea/tea/src/branch/main/Makefile#L120-L121), [Debsources](https://sources.debian.org/src/sphinx/7.2.6-4/README.rst/#L1-L3)).
